File: solutions/0/0/7/1/p0071_s0.c
/* Solution to Euler Problem 71: Ordered Fractions. */
#include "runner.h"

/* Euclidean GCD, iterative form, for reducing the candidate fraction. */
static long long gcd(long long a, long long b) {
    while (b) {
        long long t = b;
        b = a % b;
        a = t;
    }
    return a;
}

const char *solve(int argc, char *argv[]) {
    /* Farey neighbour of 3/7 in F_N: the immediate left fraction is (3k - 1)/(7k) with
       k = floor(N/7), since 3/7 - (3k-1)/(7k) = 1/(7k) is the smallest achievable gap. O(1). */
    static char _answer[32];
    long long max_d = parse_int(argv[1]);
    long long k = max_d / 7;
    /* result = 3/7 - 1/(7*k) = (3k - 1) / (7k) */
    long long num = 3 * k - 1;
    long long den = 7 * k;
    long long g = gcd(num, den);
    num /= g;
    { snprintf(_answer, sizeof _answer, "%lld", (long long)(num)); return _answer; }
}

File: solutions/public/p0071/p0071_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 71: Ordered Fractions [Level 2]. """
from __future__ import annotations

import fractions

from solver.runners import runner


@runner.main
def solve(*args: str) -> str:
    """Farey neighbour of 3/7 in F_N: the immediate left fraction is 3/7 - 1/(7k) with
    k = floor(N/7), the largest denominator multiple of 7 within the bound; O(1).
    fractions.Fraction reduces the result automatically."""
    max_d = runner.parse_int(args[0])

    result: fractions.Fraction = fractions.Fraction(3, 7) - fractions.Fraction(1, 7 * (max_d // 7))
    if runner.show:
        difference = fractions.Fraction(3, 7) - result
        print(f"Solution for max_d={max_d!r}: "
              f"result={result!r} "
              f"difference={difference!r} "
              f"result.numerator={result.numerator!r}")
    return str(result.numerator)


if __name__ == "__main__":
    raise SystemExit(solve())
